Don't publish with unsupported QoS

smurfix opened the pull request:
MQTT 3.3.1.2:
If the Server included a Maximum QoS in its CONNACK response to a Client and it receives a PUBLISH packet with a QoS greater than this, then it uses DISCONNECT with Reason Code 0x9B (QoS not supported) …

smurfix committed:
Don't publish with unsupported QoS

If the broker doesn't support QoS 1 or 2,
don't publish messages requiring it.

(Subscribing with such a QoS is still OK.)

smurfix committed:
Test all combinations of publish+subscribe QoSes

smurfix committed:
Fixed formatting.

agronholm commented:
I would be more favorable to a `maximum_qos` property.

smurfix committed:
Fix mypy error

smurfix committed:
Replace supports_qos() with a maximum_qos attribute

smurfix commented:
Well your wish is my command. *g*

agronholm commented on src/mqttproto/client_state_machine.py:
```diff
@@ -151,6 +157,7 @@ def publish(
 
         """
         self._out_require_state(MQTTClientState.CONNECTED)
+        qos = min(qos, self._maximum_qos)
```
I'm really not sure that silently downgrading the QoS is a good idea. How do other MQTT clients deal with this?

agronholm replied:
I looked at paho-mqtt and it doesn't look like they're doing this. Doing so in the state machine is even worse than in the client implementation. Exposing the `maximum_qos` information was a good call, however.

agronholm replied:
What the state machine could do is raise an error when attempting to send a message with an unsupported QoS.

smurfix replied:
Or maybe we should just drop this line and let the client app deal with the fallout instead, same as with the Retain flag handling.

agronholm replied:
Maybe, yeah, so long as we expose the maximum support QoS, this would seem reasonable.

smurfix committed:
Drop limiting QoS in `publish`

The server MAY accept those mesages even if it told us it doesn't,
thus we shouldn't prevent the client from sending them.

agronholm commented on tests/test_async_client.py:
```diff
@@ -8,12 +8,17 @@
+@pytest.mark.parametrize(
+    "qos_pub", [QoS.AT_MOST_ONCE, QoS.AT_LEAST_ONCE, QoS.EXACTLY_ONCE]
+)
+async def test_publish_subscribe(qos_sub: QoS, qos_pub: QoS) -> None:
```
What do these additions actually test for? This updated test passes just fine without the other changes in this PR.

smurfix replied:
Broker exercising. ;-)

I initially added this code to check whether it was the publish or the subscribe part that caused the problem I was seeing (ultimately fixed in #20).

I left it in because checking every possible QoS combination seems prudent, so that we don't hit any other issues along these lines.

agronholm replied:
But we're not really checking anything new here, are we? When I originally wrote this test, it wasn't passing because the I/O implementation wasn't handling the different QoS levels correctly. If these new test combinations don't add anything useful to the test suite, they should be left out. We're not trying to to validate the brokers here.

smurfix replied:
The original test didn't tell the broker which QoS level we want the messages to have, and the test suite should include that.

Alternately I could split this test in two: one that exercises every `publish` QoS level, and one that does the same with all QoS levels in `subscribe`. That'd result in running six testcases instead of nine, which seems like a somewhat marginal improvement, but it's your call.

agronholm replied:
If the intent is to test if we client can limit the maximum QoS level in its subscription, then the test should at least check the QoS level of the received messages.

smurfix replied:
OK, that's easy enough to add.

smurfix committed:
Check incoming QoS

smurfix committed:
Skip servers with smaller max-QoS

TODO switch to pytest.skip – after this and the Retain PR are merged,
because the latter contains the infrastructure to handle these pesky
BaseExceptionGroup errors.

agronholm commented on src/mqttproto/async_client.py:
```diff
@@ -480,6 +480,13 @@ async def _run_operation(self, operation: MQTTOperation[Any]) -> None:
+        """
+        Returns the maximum QoS level that the broker supports.
+        """
+        return self._state_machine.maximum_qos
```
Now we just need test coverage for this.

agronholm commented on src/mqttproto/client_state_machine.py:
```diff
@@ -161,6 +167,13 @@ def publish(
+        """
+        Returns the maximum QoS level that the broker supports.
+        """
+        return self._maximum_qos
```
This should also be explicitly covered by the client state machine tests.

agronholm commented on tests/test_async_client.py:
```diff
@@ -16,6 +16,8 @@
 async def test_publish_subscribe(qos_sub: QoS, qos_pub: QoS) -> None:
     async with AsyncMQTTClient() as client:
+        if qos_pub > client.maximum_qos:
+            return  # TODO add pytest.skip
```
Why can't you use `pytest.skip()` right now? Under what circumstances would it currently skip the test?

smurfix replied:
`pytest.skip` raises an exception which gets wrapped into an `ExceptionGroup`. Same problem as in #22 but I didn't want to replicate the solution from there here, only to then have you deal with a merge conflict.

smurfix committed:
Check that the server's QoS limit is reflected in the client

agronholm commented on tests/test_state_machines.py:
```diff
@@ -103,6 +104,25 @@ def test_client_publish_qos0(
+    client = MQTTClientStateMachine(client_id="client-X")
+    client.connect()
+    assert client.state is MQTTClientState.CONNECTING
+    _bytes = client.get_outbound_data()  # ignored
```
Is this line even needed?
```suggestion
```

smurfix replied:
Probably not in this case. I'll drop it.

agronholm commented on tests/test_state_machines.py:
```diff
@@ -103,6 +104,25 @@ def test_client_publish_qos0(
 
+@pytest.mark.parametrize("qos", [QoS.AT_MOST_ONCE, QoS.AT_LEAST_ONCE, QoS.EXACTLY_ONCE])
+def test_client_limit_qos(qos) -> None:
+    """Test that a limited QoS is processed in the client"""
```
```suggestion
    """Test that a limited QoS is processed in the client."""
```

smurfix committed:
Typo

smurfix committed:
Drop unnecessary buffer flush

agronholm commented on tests/test_state_machines.py:
```diff
@@ -103,6 +104,24 @@ def test_client_publish_qos0(
 
 
+@pytest.mark.parametrize("qos", [QoS.AT_MOST_ONCE, QoS.AT_LEAST_ONCE, QoS.EXACTLY_ONCE])
+def test_client_limit_qos(qos) -> None:
```
```suggestion
def test_client_limit_qos(qos: QoS) -> None:
```

smurfix replied:
Ah. Sure. *Sigh*

smurfix committed:
Missing type annotation

agronholm commented on tests/test_async_client.py:
```diff
@@ -8,12 +8,19 @@
-            await client.publish("test/text", "test åäö", qos=qos)
-            await client.publish("test/binary", b"\x00\xff\x00\x1f", qos=qos)
+        if qos_pub > client.maximum_qos:
+            return  # TODO add pytest.skip
```
```suggestion
            return  # TODO add pytest.skip

```

agronholm replied:
Would you mind allowing me to make changes too so I can make these trivial edits without wasting your time on them?

smurfix replied:
Well yes sure, the problem is that github doesn't even ask me whether to allow that. I haven't yet figured out where *that* setting has moved to these days, it used to be right on the create-a-PR page.

agronholm replied:
Is [this](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) helpful?

smurfix replied:
changed

smurfix replied:
> Is [this](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests) helpful?

This option isn't even displayed when I look at the PR(s).

Seems like I'll have to ask github support.

smurfix replied:
Ah. The reason was that I submitted them from my MoaT org collection instead of my personal one. *Sigh*.

No explanation why that should be so …

smurfix committed:
readability
